lib/common/variadic/uint32OrHash.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2021 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package variadic
+
+import (
+	"encoding/binary"
+	"errors"
+	"io"
+
+	"github.com/ChainSafe/gossamer/lib/common"
+)
+
+// Uint32OrHash represents a variadic type that is either uint32 or common.Hash.
+type Uint32OrHash struct {
+	value interface{}
+}
+
+// NewUint32OrHash returns a new variadic.Uint32OrHash given an int, uint32, or Hash
+func NewUint32OrHash(value interface{}) (*Uint32OrHash, error) {
+	switch v := value.(type) {
+	case int:
+		return &Uint32OrHash{
+			value: uint32(v),
+		}, nil
+	case uint32:
+		return &Uint32OrHash{
+			value: v,
+		}, nil
+	case common.Hash:
+		return &Uint32OrHash{
+			value: v,
+		}, nil
+	default:
+		return nil, errors.New("value is not uint32 or common.Hash")
+	}
+}
+
+// MustNewUint32OrHash returns a new variadic.Uint32OrHash given an int, uint32, or Hash
+// It panics if the input value is invalid
+func MustNewUint32OrHash(value interface{}) *Uint32OrHash {
+	val, err := NewUint32OrHash(value)
+	if err != nil {
+		panic(err)
+	}
+
+	return val
+}
+
+// NewUint32OrHashFromBytes returns a new variadic.Uint32OrHash from an encoded variadic uint32 or hash
+func NewUint32OrHashFromBytes(data []byte) *Uint32OrHash {
+	firstByte := data[0]
+	if firstByte == 0 {
+		return &Uint32OrHash{
+			value: common.NewHash(data[1:]),
+		}
+	} else if firstByte == 1 {
+		num := data[1:]
+		if len(num) < 4 {
+			num = common.AppendZeroes(num, 4)
+		}
+		return &Uint32OrHash{
+			value: binary.LittleEndian.Uint32(num),
+		}
+	} else {
+		return nil
+	}
+}
+
+// Value returns the interface value.
+func (x *Uint32OrHash) Value() interface{} {
+	if x == nil {
+		return nil
+	}
+	return x.value
+}
+
+// IsHash returns true if the value is a hash
+func (x *Uint32OrHash) IsHash() bool {
+	if x == nil {
+		return false
+	}
+	_, is := x.value.(common.Hash)
+	return is
+}
+
+// Hash returns the value as a common.Hash. It panics if the value is not a hash.
+func (x *Uint32OrHash) Hash() common.Hash {
+	if !x.IsHash() {
+		panic("value is not common.Hash")
+	}
+
+	return x.value.(common.Hash)
+}
+
+// IsUint32 returns true if the value is a uint32
+func (x *Uint32OrHash) IsUint32() bool {
+	if x == nil {
+		return false
+	}
+	_, is := x.value.(uint32)
+	return is
+}
+
+// Uint32 returns the value as a uint32. It panics if the value is not a uint32.
+func (x *Uint32OrHash) Uint32() uint32 {
+	if !x.IsUint32() {
+		panic("value is not uint32")
+	}
+
+	return x.value.(uint32)
+}
+
+// Encode will encode a Uint32OrHash using SCALE
```

Comment:
Is this comment still relevant? Or am i just unable to find where scale encoding happening here!

Reply:
it's encoding it using the SCALE spec, but not using the library!